keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+which we do not currently support).
+
+The controller would also remove the `TargetedByNodeMaintenance` condition from the targeted pods
+upon premature `NodeMaintenance` object removal.
+
+The controller can then show progress by reconciling `.status.targetedPods`, to show how many pods
+remain to be removed from the node.
+
+The controller would also emit a warning event if the targeted node was not marked as unschedulable.
+We could prevent this, by stomping on any changes to the node's `.spec.unschedulable` field and set
+it to true for the whole lifetime of `NodeMaintenance` object.
+
+
+### Deployment Controller
+
+The deployment controller will watch its pods and react when it observes pods with
+`TargetedByNodeMaintenance` condition. If the owning object of the pods is a deployment with a
+positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods.
+The new pods will not be scheduled on the maintained node because the `.spec.unschedulable` field
+would be set to true on that node. Once the surge pods become available, the deployment controller
+will delete the targeted pods.
+
+
+### Disruption Controller: Eviction Protection and Observability
+
+With the `.spec.MinAvailable` and `.spec.MaxUnavailable` options, PDB could allow evicting the surge
+pods because `.status.CurrentHealthy` could become higher than `.status.desiredHealthy`. This could
+disrupt ongoing migration of an application. To support mission-critical applications, we would add
+a `.spec.maxSurge` field. This would cause the disruption controller to count the
+`.status.desiredHealthy` and `.status.DisruptionsAllowed` with max surge in mind. This field could
+serve two purposes:
+
+1. If set to true, it could change the counting of status fields based on the pod workload
+   maxSurge (deployments) or based on the Surge API (see below)
+2. If set to a number (int or percentage), it would override the detected maxSurge and use only the
+   maxSurge value in the PDB object.
+
+
+```go
+type PodDisruptionBudgetSpec struct {
+    ...
+    // if true, prevents disruption of all detected surge pods
+	// if integer or percentage, prevents disruption of the defined amount of surge pods
+    MaxSurge int32|string|bool // can we do BoolOrIntOrString ?
+    ...
+}
+
+type PodDisruptionBudgetStatus struct {
+	...
+	// total number of pods that are expected to surge during a node maintenance
+    ExpectedSurgePods int32
+	...
+}
+```
+
+
+The disruption controller should check whether a workload guarded by a PDB has a pod surge capability.
+If it does, it should report `.status.expectedSurgePods` to allow cluster admins to assess which PDBs
+are blocking cluster upgrade. If the `.status.disruptionsAllowed` equals to 0, because for example
+`.spec.maxUnavailable` is equal to 0, the cluster admin can assess that the PDB will not block
+cluster upgrade if he or she observes `.status.expectedSurgePods` greater than 0.
+
+
+### Surge API (Optional)
```

Comment:
With NMO APi or a taint - workloads and controllers could watch for the maintenance request/operation.
If it can be observed, why would we need a Surge API?

Reply:
To support the PDB use case for max surge applications. Disruption controller would use this API to compute max surge resilient status for these applications.